### ISSUE TITLE
Discrepancy between paper Eq. 5 and cls_pretrain/conf.py for the stability-score numerator

### ISSUE THREAD

**Issue by xujialiu** [ChengShiest/LAST-ViT#6]
Hi, thanks for the great work and for open-sourcing the code.

While integrating LaSt-ViT into a downstream task, I noticed a small inconsistency between **Eq. 5 in the paper** and the stability-score computation in `cls_pretrain/conf.py` (class `dense_vit`). I'd like to confirm which one was used to train the released checkpoints.

### Paper (Eq. 5)

```
S[i, j] = x̂_patch[i, j] / (|x̂_patch[i, j] − x_patch[i, j]| + ε)
```

The numerator is the **low-pass-filtered** `x̂`.

### Code (`cls_pretrain/conf.py`, `dense_vit.forward`)

```python
x_detach = x[:, 1:]                                # original patches (no CLS)
x = torch.fft.fft(x[:, 1:], dim=-1)
# ... fftshift → * gaussian_kernel → ifftshift ...
x = torch.fft.ifft(x, dim=-1).real                 # x now holds the filtered x̂

diff = x_detach / torch.abs(x - x_detach)          # numerator is ORIGINAL x_detach
_, indices = torch.topk(diff, k=1, dim=1, largest=True)
sel_p = torch.gather(x_detach, 1, indices)
cls_token = torch.mean(sel_p, dim=1)
```

Here the numerator of `diff` is `x_detach` (the **original** patch features), not the filtered `x̂`. This gives

```
S[i, j] = x_patch[i, j] / (|x̂_patch[i, j] − x_patch[i, j]| + ε)
```

which differs from Eq. 5 in the numerator.

### Questions

1. Which formulation was used to train the released checkpoints (DINO / CLIP / label-supervised)?
2. Did you ablate the two variants? In my small-scale tests they agree on clearly-stable and clearly-noisy patches, but diverge in the mid-stability regime, producing different top-K selections.
3. Is one a refactoring leftover / typo, or do you consider them practically equivalent?

Happy to contribute a small clarifying note / docstring if helpful.

Thanks in advance!

**Comment by ChengShiest** [comment]
Thanks for the detailed issue and the careful comparison!

Here's the full story:

We actually tried both variants during development. Under normal top-k settings (the range we used in the paper), both formulations perform similarly — your observation is correct.

One interesting thing we found during ablation: under extreme top-k (e.g., k=1), the paper variant (numerator = filtered $\hat{x}$) exhibits what we called "gradual identity mapping": the latter half of ViT layers start outputting nearly identical features to their inputs. Counterintuitively, downstream task performance actually improves slightly in this regime. We didn't include this in the paper since we don't use extreme k values, but it's worth noting. Maybe another direction?

The code variant (numerator = original $x\_detach$) is more stable under extreme k, which is why we kept it in the released implementation.

To answer your questions:

1. Released checkpoints were trained with the code variant (x_detach as numerator).
2. We may consider including these details in a revision. Thanks for pointing this out. 

Hope this help you!
